### PR TITLE
Issue 6 - Dockerize the app

### LINKED ISSUE #6: Dockerise the app
In the previous ticket, deployed your app to Heroku using Git.

In this ticket, you will deploy your app to Heroku using [Docker](https://www.docker.com/).

## To complete this ticket, you must:

- [x] Install [Docker](https://docs.docker.com/get-started/).
- [x] Define `.dockerignore`.
- [x] Create a dedicated Docker image for your application.
  - Don't forget about JS dependencies.
- [x] Create `docker-compose.yml` file and configure multiple containers, such as the application, database and so on.
- [x] Update `database.yml` file.
- [x] Configure environment variables accordingly.

## Tips

- The process of readying your app for deployment with Docker is called [containerisation](https://www.youtube.com/watch?v=0qotVMX-J5s) (📽️ ).
- [Here](https://www.microfocus.com/documentation/enterprise-developer/ed40pu5/ETS-help/GUID-F5BDACC7-6F0E-4EBB-9F62-E0046D8CCF1B.html) (📄 ) is why Docker is a good idea.
- [A guide by Semaphore](https://semaphoreci.com/community/tutorials/dockerizing-a-ruby-on-rails-application) on how to dockerise a Rails app.

### PR DESCRIPTION
**Pull request by Applepie9**
[Issue 6](https://github.com/Applepie9/smartflix/issues/6)

To complete this ticket, you must:
- [x]  Install [Docker](https://docs.docker.com/get-started/). 
- [x]  Define .dockerignore.
- [x]  Create a dedicated Docker image for your application.
- [x] Don't forget about JS dependencies.
- [x]  Create docker-compose.yml file and configure multiple containers, such as the application, database and so on.
- [x]  Update database.yml file.
- [x]  Configure environment variables accordingly.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,19 @@
+FROM ruby:3.0.3
+
+RUN apt-get update -qq && apt-get install -y nodejs postgresql-client && apt-get install -y --no-install-recommends yarn
```
Comment (the-harry) [inline Dockerfile]: I think you can use the `apt-get` passing more than one package, something like this:

```suggestion
RUN apt-get update -qq && apt-get install -y --no-install-recommends nodejs postgresql-client yarn
```

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  db:
+    image: postgres:14.2
+    volumes:
+      - ./tmp/db:/var/lib/postgresql/data
+    environment:
+      POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: password
+  web:
+    build: .
+    command: bash -c "rm -f tmp/pids/server.pid && bundle exec rails s -p 3000 -b '0.0.0.0'"
```
Comment (the-harry) [inline docker-compose.yml]: Do we need the PID to be deleted here? Doesn't this happen already at the entry point?

File: config/database.yml [hunk 3]
```diff
@@ -3,6 +3,9 @@ default: &default
   encoding: unicode
   pool: 5
   timeout: 5000
+  host: <%= ENV['POSTGRES_HOST'] || '127.0.0.1' %>
+  username: <%= ENV['POSTGRES_USER'] || 'root' %>
+  password: <%= ENV['POSTGRES_PASSWORD'] || 'password' %>
```
Comment (the-harry) [inline config/database.yml]: Just a small suggestion:

```suggestion
  host: <%= ENV.fetch('POSTGRES_HOST') { '127.0.0.1' } %>
  username: <%= ENV.fetch('POSTGRES_USER') { 'root' } %>
  password: <%= ENV.fetch('POSTGRES_PASSWORD') { 'password' } %>
```

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    volumes:
+      - .:/smartflix
+    ports:
+      - "3000:3000"
```
Comment (the-harry) [inline docker-compose.yml]: Just another detail, but the port can be a number instead of string 💅🏻 

```suggestion
      - 3000:3000
```

**Comment by smart-jones** [comment]
@Applepie9 do you still need a review for this please?

**Comment by Applepie9** [comment]
Hi @smart-jones! Thanks for checking, I want to implement those first which @the-harry suggested. I just didn't have the time yet. After that, I would be grateful for a review 🙇🏻

**Comment by smart-jones** [comment]
> Hi @smart-jones! Thanks for checking, I want to implement those first which @the-harry suggested. I just didn't have the time yet. After that, I would be grateful for a review 🙇🏻

I'm happy to pair with you to implement these fixes sometime next week. It's totally up to you though.

**Review by the-harry — APPROVED** [review]

**Review by smart-jones — APPROVED** [review]

**Pull request merged**